### ISSUE TITLE
musing: support CREATE OR REPLACE IF UPDATED

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#6622]
In recent chats with folks using dbt + Materialize, it's come up that it would be useful for Materialize to support syntax like `CREATE OR REPLACE IF UPDATED`. 

Some background: dbt + Materialize users execute [dbt commands](https://docs.getdbt.com/reference/dbt-commands) to create materialized views in their Materialize instance. Each time the user runs these commands, like `dbt run` and `dbt seed`, we drop any previously created materialized views and create entirely new ones. This is a bit different than traditional dbt adapters, which typically just rename the old objects to keep around as back ups. (We don't support this because [Materialize can't rename materialized views](https://github.com/MaterializeInc/materialize/issues/5096).)

Users have expressed that it would be much more useful if we only dropped and recreated database objects if their definitions changed. This seems to be useful for two reasons:
1. dbt allows users to execute commands, like `dbt run`, on a model and its upstream and downstream models at one time. This is useful functionality in general, but currently blocked for those where rebuilding a view is expensive.
2. Accidentally running dbt commands more than once, which seems to happen!

If we support this, we should update our [`dbt-materialize`](https://github.com/MaterializeInc/materialize/blob/main/misc/dbt-materialize/README.md) macros!

@benesch @sploiselle @mjibson for thoughts!

**Comment by sploiselle** [comment]
Just for a little nuance here, you can renamed materialized views:

```sql
materialize=> create materialized view blah as select 1;
CREATE VIEW
materialize=> alter view blah rename to bleh;
ALTER VIEW
materialize=> select * from bleh;
 ?column?
----------
        1
(1 row)
```

...however, you cannot create another materialized view with the same name without manually renaming the index that materialized it.

Would it be possible in the DBT adapter to perform this desired index rename as part of whatever their "rename migration" path is?

**Comment by benesch** [comment]
Let's just fix #5096! I sent #6623.

**Comment by JLDLaughlin** [comment]
> Just for a little nuance here...

🙃 thank you for correcting me, @sploiselle!

> Let's just fix #5096! I sent #6623.

Awesome! I'll create an issue to update the adapter's various `CREATE` macros.

Aside from that fix, do you two (or Matt) have strong feelings either way about `CREATE OR REPLACE IF UPDATED`? While integrating the changes from #6623 will improve the adapter, it won't prevent them from creating multiple copies of potentially expensive views.

**Comment by sploiselle** [comment]
Wasn't trying to be too pedantic, sorry @JLDLaughlin!

re: `REPLACE IF UPDATED`: If you redefine a view that other views depend on, how could we propagate that re-definition?

**Comment by benesch** [comment]
Yeah, what @sploiselle brings up is the hard part! What I’m not clear on is how the `IF UPDATED` would be different from the default behavior we’d want for `CREATE OR REPLACE`. Like I think that should optimize to a no-op if you’re replacing a view with the same definition!

**Comment by JLDLaughlin** [comment]
> Wasn't trying to be too pedantic, sorry @JLDLaughlin!

No apologies necessary! The thank you was genuine, the 🙃 was at my own mistake. 

And I see both your points! I'll think about it a bit and close this if I don't come up with a reasonable proposal.
